tracer/src/Datadog.Trace/ClrProfiler/AutoInstrumentation/ManualInstrumentation/Propagators/SpanContextExtractorExtractIncludingDsmIntegration.cs

```diff
@@ -0,0 +1,47 @@
+﻿// <copyright file="SpanContextExtractorExtractIncludingDsmIntegration.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using Datadog.Trace.ClrProfiler.AutoInstrumentation.ManualInstrumentation.Proxies;
+using Datadog.Trace.ClrProfiler.CallTarget;
+using Datadog.Trace.DuckTyping;
+using Datadog.Trace.Telemetry;
+using Datadog.Trace.Telemetry.Metrics;
+
+namespace Datadog.Trace.ClrProfiler.AutoInstrumentation.ManualInstrumentation.Propagators;
+
+/// <summary>
+/// Instrumentation for <see cref="SpanContextExtractor.ExtractIncludingDsm{TCarrier}"/>
+/// </summary>
+[InstrumentMethod(
+    AssemblyName = "Datadog.Trace.Manual",
+    TypeName = "Datadog.Trace.SpanContextExtractor",
+    MethodName = "ExtractIncludingDsm",
+    ReturnTypeName = "Datadog.Trace.ISpanContext",
+    ParameterTypeNames = ["!!0", "System.Func`3[!!0,System.String,System.Collections.Generic.IEnumerable`1[System.String]]", "System.String", "System.String"],
+    MinimumVersion = ManualInstrumentationConstants.MinVersion,
+    MaximumVersion = ManualInstrumentationConstants.MaxVersion,
+    IntegrationName = ManualInstrumentationConstants.IntegrationName)]
+[Browsable(browsable: false)]
+[EditorBrowsable(EditorBrowsableState.Never)]
+public class SpanContextExtractorExtractIncludingDsmIntegration
+{
+    internal static CallTargetState OnMethodBegin<TTarget, TCarrier, TAction>(TTarget instance, in TCarrier carrier, in TAction getter, string messageType, string source)
+    {
+        TelemetryFactory.Metrics.Record(PublicApiUsage.SpanContextExtractor_ExtractIncludingDsm);
+        var extract = (Func<TCarrier, string, IEnumerable<string?>>)(object)getter!;
```

Comment:
Although they're not annotated like that, the carrier, getter, messageType, and source _could_ all be null as they're provided by the customer. Maybe we should check them for null (after recording the metric) and throw `ArgumentNullException`?

Reply:
there is this code
https://github.com/DataDog/dd-trace-dotnet/blob/4157f1a3f9288124b5bbe3b31b987a3fa53ce077/tracer/src/Datadog.Trace/Propagators/SpanContextPropagator.cs#L145-L146
which is a bit deeper in the call stack, but does the job I'd say :)

Reply:
What could be handled better is the case where messageType and source are not the same null if you know what I mean. I'll do that.